chefkoch/recipe.py

```diff
@@ -0,0 +1,771 @@
+                if output in invertedInputDict:
+                    logger.debug("Taking the edge from " + node.name + " to " +
+                                 nextNode.name)
+                    if self.recursiveDFS(nextNode, nodesOnTheWay):
+                        return True # a circle was found
+                    # else continue with next node
+        # after all outgoing edges where tested, remove current node
+        nodesOnTheWay.remove(node)
+        # if there is no circle from none of the outputs
+        return False
+
+
+class Node:
+    """
+    A node encapsules a simulation step within a recipe. The step can
+    be realised by a python file, a sub-recipe or a built-in function.
+    Each node also has a name, a dict of inputs and a dict of outputs.
+    To the inputdict, the key is the same input name that the step takes,
+    the value is where the input comes from. To the outputdict, the key
+    is the name the step uses, the value is the name under which the
+    output is available to other nodes in the recipe (the same name used
+    as value in another inputdict).
+    """
+    def __init__(self, name, inputdict, outputdict, stepsource):
+        """
+        Initializes a node of the recipe. A node represents a simulation
+        step.
+
+        :param name: Name of the simulation step.
+        :type name: string or unicode
+        :param inputdict: Dictionary of all inputs needed to execute this step.
+        :type inputdict: dictionary
+        :param outputdict: Dictionary of all outputs of the simulation step.
+        :type outputdict: dictionary
+        :param stepsource: Information on how to execute this step.
+        :type stepsource: string
+        """
+        # for empty name enter "" into recipe
+        # unicode and string needed
+        try:
+            name_obj = Name(name)
+            self.name = name_obj.name # Willi, ist das wirklich so gemeint?
+        except TypeError as err:
+            pass
+        # testing the input to be delivered in a dict
+        if not (isinstance(inputdict, dict)):
+            raise TypeError('The input of node ' + str(name) +
+                            ' must be of the format {\"name as in ' +
+                            'step\": value, ...}')
+            return
+        self.inputs = inputdict
+        # later replace strings by values in flavour?
+        # testing the output to be delivered in a dict
+        if not (type(outputdict) == dict):
+            raise TypeError('The output of node ' + str(name) +
+                            ' must be of the format {\"name as in ' +
+                            'step\": value, ...}')
+            return
+        self.outputs = outputdict
+        try:
+            step_obj = StepSource(stepsource)
+            self.step = step_obj.step
+        except TypeError as err:
+            pass # todo abort and throw warning/error, ignore whole node
+
+
+class Name:
```

Comment:
Always specify a parent!
```
class Name(object):
```
Otherwise certain features of python3 will break, IIRC `super`

Reply:
Was ist denn der Parent einer neuen Klasse, z.B. der Name Klasse?

Reply:
Danke, ich merke es mir. Aber bis jetzt passt der Code denke ich so, da Name von nichts erbt bzw. keine Parent Klasse hat.
https://www.pythonforbeginners.com/super/working-python-super-function